### PR TITLE
Overlap gradient computation and NCCL AllReduce

### PR DESCRIPTION
**Pull request by PeterZhizhin**
On my setup, I get the following:

Before:
```
step    2/37: train loss 4.720275 (acc 4.688650) (224.046844 ms, 36563.773438 tok/s)
step    3/37: train loss 3.802741 (acc 3.943135) (224.151611 ms, 36555.007812 tok/s)
step    4/37: train loss 3.698719 (acc 3.800745) (227.287033 ms, 36375.347656 tok/s)
step    5/37: train loss 3.444999 (acc 3.528596) (227.886978 ms, 36260.062500 tok/s)
```

After:
```
step    2/37: train loss 4.715888 (acc 4.686493) (199.011169 ms, 41163.503906 tok/s)
step    3/37: train loss 3.798963 (acc 3.942383) (193.084412 ms, 41811.468750 tok/s)
step    4/37: train loss 3.697987 (acc 3.800879) (193.079300 ms, 42027.660156 tok/s)
step    5/37: train loss 3.444056 (acc 3.526504) (193.470459 ms, 42112.496094 tok/s)
```

So, a 12% speedup.

NSight Systems profiles:

Before:
NSight Compute profile before: backward then NCCL happen on the same stream

After:
NSight Compute profile before: backward and NCCL is overlapped

### CODE REVIEW

File: train_gpt2.cu [hunk 1]
```diff
@@ -2348,7 +2410,7 @@ void common_free(GPT2 &model) {
     cudaCheck(cudaFree(cublaslt_workspace));
     cublasCheck(cublasDestroy(cublas_handle));
     cublasCheck(cublasLtDestroy(cublaslt_handle));
-    create_cudnn();
+    destroy_cudnn();
```
Comment (ngc92) [inline train_gpt2.cu:2413]: @karpathy @PeterZhizhin cherry pick; this should be merged immediately

File: train_gpt2.cu [hunk 2]
```diff
@@ -2736,14 +2795,14 @@ int main(int argc, char *argv[]) {
...
-        float accumulated_loss = multi_gpu_config.num_processes == 1 ? model.mean_loss : model.accumulated_mean_loss;
+        float accumulated_loss = multi_gpu_cpu_float_mean(model.mean_loss, &multi_gpu_config);
         printf0("step %4d/%d: train loss %f (acc %f) (%f ms, %0f tok/s)\n",
                 step + 1, train_num_batches, model.mean_loss, accumulated_loss,
                 time_elapsed_ms, bias_corrected_ema_tokens_per_second);
         logger_log_train(&logger, step, model.mean_loss);
 
         // disable the profiler after 3 steps of optimization
-        if (step == 3) { cudaProfilerStop(); }
+        if (step == 3) { cudaCheck(cudaProfilerStop()); }
```
Comment (ngc92) [inline train_gpt2.cu:2805]: this isan independent fix too

File: train_gpt2.cu [hunk 3]
```diff
@@ -2147,6 +2190,11 @@ void gpt2_backward(GPT2 *model) {
     floatX* dresidual = (floatX*)grads_acts.residual3; // the main buffer holding the gradient in the backward pass
     layernorm_backward(dresidual, grads.lnfw, grads.lnfb, scratchF, grads_acts.bt4c, residual, params.lnfw, acts.lnf_mean, acts.lnf_rstd, B, T, C);
 
+    // Aggregate grads.lnfw and grads.lnfb in a background stream
+    floatX* layernorm_backward_pointers[] = {grads.lnfw, grads.lnfb};
+    size_t layernorm_backward_sizes[] = {C, C};
+    multi_gpu_async_all_reduce_pointers_group(2, layernorm_backward_pointers, layernorm_backward_sizes, multi_gpu_config, main_stream);
```
Comment (ngc92) [inline train_gpt2.cu:2196]: comment  says background stream, but  call uses main_stream?

Comment (ngc92) [reply train_gpt2.cu:2196]: oh wait, in the version this code was  based on, main_stream was the background stream?
